### ISSUE TITLE
Karma Issues - Running Angular Unit Tests

### ISSUE THREAD

**Issue by FA-tvonmoll** [GoogleCloudPlatform/cloud-builders#232]
Greetings!

TL;DR: being able to run Angular unit tests with karma on cloud builder would be brilliant

I absolutely love the idea of cloud builder (no need to maintain a separate build cluster and painless pushes to the container registry), but I've hit a major blocker. For our frontend projects, we've been using [Angular 5](https://angular.io). The problem is that karma requires a browser (even when just running the unit tests), and we have yet to find a way to accommodate that in cloud builder.

The closest I've gotten is with `FirefoxHeadless`, but I've also tried the ChromeHeadless/Puppeteer route to no avail (the latter is also a bit more painful/time-consuming to install onto the build container each time).

Here's an example `cloudbuild.yaml` in a new angular project:

```
steps:
- name: 'gcr.io/cloud-builders/npm'
  args: ['install', 'karma-firefox-launcher']
- name: 'gcr.io/cloud-builders/npm'
  args: ['install']
- name: 'gcr.io/cloud-builders/npm'
  env: ['NO_PROXY=localhost,0.0.0.0/4201,0.0.0.0/9876']
  args: ['test', '--', '--singleRun', '--browsers=FirefoxHeadless']
- name: 'gcr.io/cloud-builders/docker'
  args: ['build', '-t', 'gcr.io/$PROJECT_ID/karma-test', '.']
images: ['gcr.io/$PROJECT_ID/karma-test']
```

And I've been testing locally with:

```
container-builder-local --config=cloudbuild.yaml --dryrun=false  .
```

The buck stops here:

```
Step #3: Already have image (with digest): gcr.io/cloud-builders/npm
Starting Step #3
Step #3:
Step #3: > karma-test-ng@0.0.0 test /workspace
Step #3: > ng test "--single-run" "--browsers=FirefoxHeadless"
Step #3:
Step #3: 15 03 2018 11:51:05.323:INFO [karma]: Karma v2.0.0 server started at http://0.0.0.0:9876/
Step #3: 15 03 2018 11:51:05.327:INFO [launcher]: Launching browser FirefoxHeadless with unlimited concurrency
Step #3: 15 03 2018 11:51:05.335:INFO [launcher]: Starting browser FirefoxHeadless
Step #3: 15 03 2018 11:54:35.381:WARN [launcher]: FirefoxHeadless have not captured in 210000 ms, killing.
Step #3: 15 03 2018 11:54:37.390:WARN [launcher]: FirefoxHeadless was not killed in 2000 ms, sending SIGKILL.
Step #3: 15 03 2018 11:54:39.395:WARN [launcher]: FirefoxHeadless was not killed by SIGKILL in 2000 ms, continuing.
Step #3: npm ERR! Test failed.  See above for more details.
Finished Step #3
2018/03/15 07:54:37 status changed to "ERROR"
```

Thanks!

**Comment by bendory** [comment]
I know nothing about karma (and I'm not sure that anyone on the GCB team does either). Question: leaving GCB completely out of the equation (including leaving container-builder-local out of the equation), are you able to execute the above using `docker run`?

I can't play with this without (at least) knowing what's in your package.json. To share confidential information, send an email to gcr-contact@google.com and link to this issue...

**Comment by FA-tvonmoll** [comment]
Thanks for your quick reply, @bendory !

I'll put together a repo for this (so far, I'm able to reproduce with a new angular project from cli).

Yes, I was able to build a plain docker image that runs the tests. I reckon I could do a multiphase Dockerfile (and just run the tests as part of the `docker build` step), but I would lose the nice steps on the output side in the build history. I'll include that Dockerfile and the cloudbuild.yaml file in the reference repo whenever I get a few minutes to get it out there.

**Comment by FA-tvonmoll** [comment]
[Reference repo](https://github.com/FA-tvonmoll/gcb-angular-test).

Slightly different cloudconfig than originally reported, but you get the idea.

`$ docker build -t gcb-angular-test .` --> The three tests added by the angular cli run in the container.

`$ container-builder-local --config=cloudbuild.yaml --dryrun=false  .` same error as originally reported.

My goal was to translate as directly as possible the dockerfile over to build steps, but it's definitely possible that I'm missing something obvious.

Thanks!

**Comment by FA-tvonmoll** [comment]
Ah, I was pretty far off in what I thought that `ubuntu` block accomplished.

Thanks for those suggestions and clarification! Just to make sure I'm not further straying from the appropriate course: 

I could adapt that to more of a pipeline by:

1. Building an image that has the source and FF (call it `Dockerfile.unitTests` or something)
2. Pushing it 
3. Calling that step from your `runtest.yaml` example
4. and from there, could continue to do whatever else needs to happen to make the final image with the compiled app + nginx.

Something like:

```
steps:
- name: 'gcr.io/cloud-builders/docker'
  args: ['build', '-t', 'gcr.io/$PROJECT_ID/gcb-angular-test-uts', '-f', 'Dockerfile.unitTests', '.']
- name: 'gcr.io/cloud-builders/docker'
  args: ["push", "gcr.io/$PROJECT_ID/gcb-angular-test-uts"]
- name: 'gcr.io/$PROJECT_ID/gcb-angular-test'
  env: ['NO_PROXY=0.0.0.0/9876', 'FIREFOX_BIN=/usr/bin/firefox']
  args: ['npm', 'test', '--', '--singleRun', '--browsers=FirefoxHeadless']
```

**Comment by bendory** [comment]
You could do it as you describe -- but note that if the only purpose of the `gcb-angular-test-uts` is to run the unit test, you need not actually push it. Any container that you build in a prior build step is available for execution in a subsequent step, so I can reduce your example to:

```
steps:
- name: 'gcr.io/cloud-builders/docker'
  args: ['build', '-t', 'gcr.io/$PROJECT_ID/gcb-angular-test-uts', '-f', 'Dockerfile.unitTests', '.']
- name: 'gcr.io/$PROJECT_ID/gcb-angular-test-uts'
  env: ['NO_PROXY=0.0.0.0/9876', 'FIREFOX_BIN=/usr/bin/firefox']
  args: ['npm', 'test', '--', '--singleRun', '--browsers=FirefoxHeadless']
```

Note that if this is my entire build configuration, the `gcr.io/$PROJECT_ID/gcb-angular-test-uts ` container is discarded once the test completes. A subsequent step could push it or you could list it in the `images` field and we'll push it for you.

I'm closing this thread because the initially reported issue is resolved; feel free to reopen if you have additional questions.
